// marko/dist/components/ServerComponent.js
'use strict';

class ServerComponent {
    constructor(id, input, out, typeName, customEvents, scope) {
        this.id = id;
        this.w_ = customEvents;
        this.k_ = scope;
        this._F_ = undefined;
        this.y_ = undefined;
        this.r_ = undefined;
        this.typeName = typeName;
        this.v_ = undefined; // Used to keep track of bubbling DOM events for components rendered on the server
        this._G_ = 0;

        if (this.onCreate !== undefined) {
            this.onCreate(input, out);
        }

        if (this.onInput !== undefined) {
            var updatedInput = this.onInput(input, out) || input;

            if (this.y_ === undefined) {
                this.y_ = updatedInput;
            }

            this._F_ = updatedInput;
        } else {
            this.y_ = this._F_ = input;
        }

        if (this.onRender !== undefined) {
            this.onRender(out);
        }
    }

    set input(newInput) {
        this.y_ = newInput;
    }

    get input() {
        return this.y_;
    }

    set state(newState) {
        this.r_ = newState;
    }

    get state() {
        return this.r_;
    }

    get _e_() {
        return this.r_;
    }

    elId(scopedId, index) {
        var id = this.id;

        var elId = scopedId != null ? id + "-" + scopedId : id;

        if (index != null) {
            elId += "[" + index + "]";
        }

        return elId;
    }
}

ServerComponent.prototype.getElId = ServerComponent.prototype.elId;

module.exports = ServerComponent;
