// marko/dist/components/State.js
var extend = require('raptor-util/extend');

function ensure(state, propertyName) {
    var proto = state.constructor.prototype;
    if (!(propertyName in proto)) {
        Object.defineProperty(proto, propertyName, {
            get: function () {
                return this._f_[propertyName];
            },
            set: function (value) {
                this.N_(propertyName, value, false /* ensure:false */);
            }
        });
    }
}

function State(component) {
    this.h_ = component;
    this._f_ = {};

    this.D_ = false;
    this.S_ = null;
    this.R_ = null;
    this._H_ = null; // An object that we use to keep tracking of state properties that were forced to be dirty

    Object.seal(this);
}

State.prototype = {
    l_: function () {
        var self = this;

        self.D_ = false;
        self.S_ = null;
        self.R_ = null;
        self._H_ = null;
    },

    L_: function (newState) {
        var state = this;
        var key;

        var rawState = this._f_;

        for (key in rawState) {
            if (!(key in newState)) {
                state.N_(key, undefined, false /* ensure:false */, false /* forceDirty:false */);
            }
        }

        for (key in newState) {
            state.N_(key, newState[key], true /* ensure:true */, false /* forceDirty:false */);
        }
    },
    N_: function (name, value, shouldEnsure, forceDirty) {
        var rawState = this._f_;

        if (shouldEnsure) {
            ensure(this, name);
        }

        if (forceDirty) {
            var forcedDirtyState = this._H_ || (this._H_ = {});
            forcedDirtyState[name] = true;
        } else if (rawState[name] === value) {
            return;
        }

        if (!this.D_) {
            // This is the first time we are modifying the component state
            // so introduce some properties to do some tracking of
            // changes to the state
            this.D_ = true; // Mark the component state as dirty (i.e. modified)
            this.S_ = rawState;
            this._f_ = rawState = extend({}, rawState);
            this.R_ = {};
            this.h_.M_();
        }

        this.R_[name] = value;

        if (value === undefined) {
            // Don't store state properties with an undefined or null value
            delete rawState[name];
        } else {
            // Otherwise, store the new value in the component state
            rawState[name] = value;
        }
    },
    toJSON: function () {
        return this._f_;
    }
};

module.exports = State;
